File: original/tetris.c

#include "tetris.h"

#define		FOR(i, len)		for(i = 0; i < len; i++) // Shortcut for loop 

 // Global constant for the number of cordinates in a Tetrimino
static const int CORD_NUM = 3; // Doesn't count the center
static const int B_LEN = 10;
static const int B_HEI = 22;

static const int O_CORDS[3][2] = 
{
	{1, 0},
	{0, 1},
	{1, 1}
};

static const int I_CORDS[3][2] = 
{
	{-1, 0},
	{1, 0},
	{2, 0}
};

static const int S_CORDS[3][2] = 
{
	{-1, 0},
	{0, 1},
	{1, 1}
};

static const int Z_CORDS[3][2] = 
{
	{1, 0},
	{0, 1},
	{-1, 1}
};

static const int L_CORDS[3][2] = 
{
	{-1, 0},
	{1, 0},
	{1, 1}
};

static const int J_CORDS[3][2] = 
{
	{1, 0},
	{-1, 0},
	{-1, 1}
};

static const int T_CORDS[3][2] = 
{
	{-1, 0},
	{0, 1},
	{1, 0}
};

int CORD_LIST[7][3][2]; 

static void 
init_cords(void)
{
	memcpy(&CORD_LIST[0], &O_CORDS, sizeof O_CORDS); 
	memcpy(&CORD_LIST[1], &I_CORDS, sizeof I_CORDS); 
	memcpy(&CORD_LIST[2], &S_CORDS, sizeof S_CORDS);
	memcpy(&CORD_LIST[3], &Z_CORDS, sizeof Z_CORDS);
	memcpy(&CORD_LIST[4], &L_CORDS, sizeof L_CORDS);
	memcpy(&CORD_LIST[5], &J_CORDS, sizeof J_CORDS);
	memcpy(&CORD_LIST[6], &T_CORDS, sizeof T_CORDS);
}

// Creator of my structures
mino 
create_mino(mino *piece)
{
	static int call_num;
	if(!call_num) 
	{
		init_cords();
		call_num++;
	}

	int cent_x = 5;
	int cent_y = 20;

	piece->shape = random() % 7 + 1;
	piece->ori = ori_D;
	
	piece->center[0] = cent_x;
	piece->center[1] = cent_y; 
	
	int i;
	FOR(i, CORD_NUM)
	{
		piece->cords[i][0] = 
			cent_x + CORD_LIST[piece->shape -1 ][i][0];
		
		piece->cords[i][1] = 
			cent_y + CORD_LIST[piece->shape -1 ][i][1];
	}

	return *piece;
}

game 
create_game(game *tetris)
{
	int i, j;
	FOR(i, B_LEN)
	{
		FOR(j, B_HEI)
		{
			tetris->board[i][j].state = empty;
		}
	}

	tetris->curr_mino = 
		create_mino(&tetris->curr_mino);

	FOR(i, CORD_NUM)
	{
		tetris->next_minos[i] = 
			create_mino(&tetris->next_minos[i]);
	}

	tetris->level = 1;
	tetris->drop_speed = 950; 

	tetris->lines = 0;
	tetris->time = 0;
	tetris->score = 0;

	return *tetris;

}


// Turns the piece 90 degree clockwise around its center index
// I shape changes the center index and O shape can't rotate
void 
rotate(mino *piece)
{
	int cent_x = piece->center[0];
	int cent_y = piece->center[1];
	
	int i;
	FOR(i, CORD_NUM)
	{	
		int curr_x = piece->cords[i][0];
		int curr_y = piece->cords[i][1];

		piece->cords[i][0] = cent_x + (curr_y - cent_y);
		piece->cords[i][1] = cent_y - (curr_x - cent_x);

	}
	
	piece->ori = ++piece->ori % 4;
	
	if(piece->shape == shape_I && piece->ori % 2)
	{
		int temp_y = piece->center[1];
		piece->center[1] = piece->cords[1][1];
		piece->cords[1][1] = temp_y;
	}
}

int  
can_rotate(game *tetris)
{
	mino *piece = &tetris->curr_mino;

	if(piece->shape == shape_O) return 0;

	int cent_x = piece->center[0];
	int cent_y = piece->center[1];
	
	int i;
	FOR(i, CORD_NUM)
	{	
		int curr_x = piece->cords[i][0];
		int curr_y = piece->cords[i][1];

		int new_x = cent_x + (curr_y - cent_y);
		int new_y = cent_y - (curr_x - cent_x);

		if(new_x >= B_LEN || new_x < 0 || new_y < 0)
		{
			return 0;	
		}

		else if(tetris->board[new_x][new_y].state == placed)
		{
			return 0;
		}
	}
	
	return 1;
}

// Moves the piece left or right one based on the value of direction being + or -
void 
shift(mino *piece, int dir)
{	
	if(!(dir * dir - 1))
	{	
		piece->center[0] += dir; 

		int i;
		FOR(i, CORD_NUM)
		{
			piece->cords[i][0] += dir;
		}
	}
}


int
can_shift(game *tetris, int dir)
{
	if(dir * dir - 1) return 0;

	mino *piece = &tetris->curr_mino;
	
	int cx = piece->center[0];
	int cy = piece->center[1];

	if( cx + dir >= B_LEN || cx + dir < 0 || 
			tetris->board[cx + dir][cy].state == placed)
	{
		return 0;
	}
	
	int i;
	FOR(i, CORD_NUM)
	{
		int x = piece->cords[i][0];
		int y = piece->cords[i][1];

		if( x + dir >= B_LEN || x + dir < 0 || 
				tetris->board[x+dir][y].state == placed)
		{
			return 0;
		}
	}
	
	return 1;
}


// Moves the piece down one
void 
fall(mino *piece)
{
	piece->center[1] -= 1; 

	int i;
	FOR(i, CORD_NUM)
	{
		piece->cords[i][1] -= 1;
	}

}	

int
can_fall(game *tetris)
{
	mino *piece = &tetris->curr_mino;
	
	int cx = piece->center[0];
	int cy = piece->center[1];

	if( cy - 1 < 0 || tetris->board[cx][cy-1].state == placed)
	{
		return 0;
	}
	
	int i;
	FOR(i, CORD_NUM)
	{
		int x = piece->cords[i][0];
		int y = piece->cords[i][1];
		
		if(x == cx && y > cy)
		{
			continue;
		}

		else if( y - 1 < 0 || tetris->board[x][y-1].state == placed)
		{
			return 0;
		}
	}
	
	return 1;
}


// The updates the location of the current tetrimino 
void 
update_pos(game *tetris)
{
	int drop = drop_distance(tetris);
	mino *piece = &tetris->curr_mino;

	int i, j;
	FOR(i, B_LEN)
	{
		FOR(j, B_HEI)
		{
			if(tetris->board[i][j].state == falling || 
					tetris->board[i][j].state == phantom)
			{
				tetris->board[i][j].state = empty;
				tetris->board[i][j].color = empty;
			}
		}
	}
	
	int x = piece->center[0];
	int y = piece->center[1];

	tetris->board[x][y - drop].state = phantom;
	tetris->board[x][y].state = falling;
	tetris->board[x][y].color = piece->shape;


	FOR(i, CORD_NUM)
	{
		x = piece->cords[i][0];
		y = piece->cords[i][1];
	
		tetris->board[x][y - drop].state = phantom;
		tetris->board[x][y].state = falling;
		tetris->board[x][y].color = piece->shape;
	}
}


void
place(game *tetris)
{
	int x = tetris->curr_mino.center[0];
	int y = tetris->curr_mino.center[1];

	tetris->board[x][y].state = placed;
	
	int i;
	FOR(i, CORD_NUM)
	{
		x = tetris->curr_mino.cords[i][0];
		y = tetris->curr_mino.cords[i][1];

		tetris->board[x][y].state = placed;
	}
}



// Releases the next mino making it the current Tetrimino, in the invisible 
// rows of the board (above row 20)
void 
release_next(game *tetris)
{
	tetris->curr_mino = 
		tetris->next_minos[0];
	
	tetris->next_minos[0] = 
		tetris->next_minos[1];
	
	tetris->next_minos[1] = 
		tetris->next_minos[2];
	
	tetris->next_minos[2] = 
		create_mino(&tetris->next_minos[2]);
}


// Gets the distance the current Tetrimino will drop if dropped
int
drop_distance(game *tetris)
{	
	mino piece = tetris->curr_mino;
	
	int i, j, min_move = B_HEI;

	int cx = piece.center[0];
	int cy = piece.center[1];

	for(j = cy; j >= 0; j--)
    {
        if(tetris->board[cx][j].state == placed)
        {
            if(cy -j -1 < min_move)
            {
                min_move = cy -j -1;
            }
			break;
        }

		else if(j == 0 && cy < min_move)
		{
			min_move = cy;
		}
    }

	FOR(i, CORD_NUM)
	{
		int x = piece.cords[i][0];
		int y = piece.cords[i][1];

		if (x != cx || y < cy)
		{
			for(j = y; j >= 0; j--)
			{
				if(tetris->board[x][j].state == placed)
				{
					if(y -j -1 < min_move)
					{
						min_move = y -j -1;
					}
					
					break;
				}

				else if(j == 0 && y < min_move)
				{
					min_move = y;
				}
			}
		}
	}

	return min_move;
}


// Drops the current Tetrimino on top of the pieces below it
void 
drop(game *tetris)
{
	int dist = drop_distance(tetris);
	mino *piece = &tetris->curr_mino;

	piece->center[1] -= dist;
	
	int i;
	FOR(i, CORD_NUM)
	{
		piece->cords[i][1] -= dist;
	}

	update_pos(tetris);
}


// Stores the current Tetrimino and releases the previously held Tetrimino 
// if there is one. Otherwise releases the next Tetrimino 
void 
hold(game *tetris)
{
	int cent_x = 5;
	int cent_y = 20;

	mino *piece = &tetris->held_mino;

	static int count;

	if(count)
	{
		mino temp = tetris->curr_mino;
		tetris->curr_mino = tetris->held_mino;
		tetris->held_mino = temp;
	}
	
	else
	{
		tetris->held_mino = tetris->curr_mino;
		release_next(tetris);
		++count;
	}	

	piece->center[0] = cent_x;
	piece->center[1] = cent_y; 
	
	int i;
	FOR(i, CORD_NUM)
	{
		piece->cords[i][0] = 
			cent_x + CORD_LIST[piece->shape -1 ][i][0];
		
		piece->cords[i][1] = 
			cent_y + CORD_LIST[piece->shape -1 ][i][1];
	}

	update_pos(tetris);
}


static void
shift_all_down(game *tetris, int bottom)
{
	int i, j;
	for(j = bottom; j < B_HEI -1; j++)
	{
		FOR(i, B_LEN)
		{
			tetris->board[i][j] = tetris->board[i][j+1];
		}
	}
}


// Reads the possible rows to clear and if they are full clears
void
line_up(game *tetris)
{
	int i, j; 
	int count, num_rows = 0;
	
	FOR(j, B_HEI)
	{	
		count = 0;

		FOR(i, B_LEN)
		{
			count += tetris->board[i][j].state;
		}
		
		if(count == B_LEN * placed)
		{
			shift_all_down(tetris, j);

			tetris->lines++;
			num_rows++;
			j--;
		}
		
		if(num_rows == 4) break;
	}

	if(tetris->lines / ( 5 * ( tetris->level + 1 ) ) )
	{
		tetris->level++;
		tetris->lines = 0;
		calc_speed(tetris);
	}
}


// Calculates the speed based on what the current level is
void
calc_speed(game *tetris)
{	
	int speed = 950;
	int level = tetris->level;
	tetris->drop_speed = speed * (.95 - (level -1) / level);
}

File: src/tetris.rs
pub use self::tetris::Game;

mod tetris {
    use rand::{thread_rng, Rng};
    use std::cmp::min;
    
    // Private Data
    
    // Global constant for the number of cordinates in a Tetrimino
    const CORD_NUM: usize = 3; // Doesn't count the center
    const NUM_MINOS: usize = 3; // Number of held minos
    const B_LEN: usize = 10; // Board Sizes (Length and Height)
    const B_HEI: usize = 22;

    const CORD_LIST: [[[i8; 2]; 3]; 7] =  
    // Shape O
    [
        [
            [1, 0],
            [0, 1],
            [1, 1]
        ], 
        
        // Shape I
        [
            [-1, 0],
            [1, 0],
            [2, 0]
        ],
        
        // Shape S
        [
            [-1, 0],
            [0, 1],
            [1, 1]
        ],
        
        // Shape Z
        [
            [1, 0],
            [0, 1],
            [-1, 1]
        ], 
        
        // Shape L
        [
            [-1, 0],
            [1, 0],
            [1, 1]
        ],
        
        // Shape J
        [
            [1, 0],
            [-1, 0],
            [-1, 1]
        ],
        
        // Shape T
        [
            [-1, 0],
            [0, 1],
            [1, 0]
        ]
    ];

    const BLACK: usize = 7;

    const COLOR_LIST: [[u8; 3]; 8] = [
        [236, 217, 50], // Shape O = Yellow
        [235, 128, 32], // Shape I = Orange
        [82, 224, 82],  // Shape S = Green
        [229, 68, 46],  // Shape Z = Red
        [42, 255, 223], // Shape L = Cyan
        [34, 142, 255], // Shape J = Blue
        [186, 35, 230], // Shape T = Purple
        [0, 0, 0],   // Black
    ];

    const MAX_LEVEL : usize = 19;
    const DROP_TICKS : [u16; MAX_LEVEL+1] = [  
      // 0,  1,  2,  3,  4,  5,  6,  7,  8,  9,
        30, 28, 26, 24, 22, 20, 18, 16, 14, 12,
      //10, 11, 12, 13, 14, 15, 16, 17, 18, 19
        10,  9,  8,  7,  6,  5,  4,  3,  2,  1
    ];


    // Possible shapes
    #[derive(Copy, Clone)]
    enum Shape {
        ShapeO,
        ShapeI,
        ShapeS,
        ShapeZ,
        ShapeL,
        ShapeJ,
        ShapeT,
    }

    impl Default for Shape {
        fn default() -> Self {
            Shape::ShapeO
        }
    }

    // Possible orientations Up, Left, Right and Down
    #[derive(Copy, Clone)]
    enum Orientation {
        OriU,
        OriR,
        OriD,
        OriL,
    }

    impl Default for Orientation {
        fn default() -> Self {
            Orientation::OriU
        }
    }

    // Struct of a Tetrimino
    #[derive(Default, Copy, Clone)]
    struct Mino {
        // Ints to hold the values of the shape and orientation
        // Values are only to be thoses in the corresponding enum
        shape: Shape,
        ori: Orientation,

        // The center cordinate of the Tetrimino
        // Is used as the origin of rotation
        // I shape swaps its cordinate depending on the orientation
        center: [u8; 2],

        // 2D array that holds the cordinates of spaces the Tetrimino is in
        // except for the center cord
        cords: [[u8; 2]; 3],
    }

    impl Mino {
        // Creator of my structures
        fn new() -> Mino {
            let mut piece: Mino = Default::default();

            let cent_x: i8 = 5;
            let cent_y: i8 = 20;

            let mut rng = thread_rng();
            let mut rand: u32 = rng.gen();
            rand = rand % 7;
            piece.shape = match rand {
                1 => Shape::ShapeI,
                2 => Shape::ShapeS,
                3 => Shape::ShapeZ,
                4 => Shape::ShapeL,
                5 => Shape::ShapeJ,
                6 => Shape::ShapeT,
                _ => Shape::ShapeO,
            };

            piece.ori = Orientation::OriD;

            piece.center[0] = cent_x as u8;
            piece.center[1] = cent_y as u8;

            for i in 0..CORD_NUM {
                piece.cords[i][0] = (cent_x + CORD_LIST[piece.shape as usize][i][0]) as u8;

                piece.cords[i][1] = (cent_y + CORD_LIST[piece.shape as usize][i][1]) as u8;
            }

            return piece;
        }

        // Turns the piece 90 degree clockwise around its center index
        // I shape changes the center index and O shape can't rotate
        fn rotate(&mut self) {
            let cent_x = self.center[0] as i16;
            let cent_y = self.center[1] as i16;

            for i in 0..CORD_NUM {
                let curr_x = self.cords[i][0] as i16;
                let curr_y = self.cords[i][1] as i16;

                self.cords[i][0] = (cent_x + (curr_y - cent_y)) as u8;
                self.cords[i][1] = (cent_y - (curr_x - cent_x)) as u8;
            }

            self.ori = match self.ori {
                Orientation::OriU => Orientation::OriR,
                Orientation::OriR => Orientation::OriD,
                Orientation::OriD => Orientation::OriL,
                Orientation::OriL => Orientation::OriU,
            };

            match self.shape {
                Shape::ShapeI => {
                    if self.ori as i32 % 2 == 0 {
                        let temp_y = self.center[1];
                        self.center[1] = self.cords[1][1];
                        self.cords[1][1] = temp_y;
                    }
                }
                _ => {}
            }
        }

        // Moves the piece left or right one with false being left and true being right
        fn shift(&mut self, dir: bool) {
            if dir {
                self.center[0] += 1;
            } else {
                self.center[0] -= 1;
            }

            for i in 0..CORD_NUM {
                if dir {
                    self.cords[i][0] += 1;
                } else {
                    self.cords[i][0] -= 1;
                }
            }
        }

        // Moves the piece down one
        fn fall(&mut self) {
            self.center[1] -= 1;

            for i in 0..CORD_NUM {
                self.cords[i][1] -= 1;
            }
        }
    }

    // Possible States of Block in Game Board
    #[derive(PartialEq, Copy, Clone)]
    enum State {
        Empty,
        Falling,
        Placed,
        Phantom,
    }

    impl Default for State {
        fn default() -> Self {
            State::Empty
        }
    }

    // Public Data
    #[derive(Default, Copy, Clone)]
    pub struct Block {
        state: State,
        pub color: [u8; 3],
    }

    // Struct of the Game
    // Holds the info of everything used in the game of Tetris
    #[derive(Default)]
    pub struct Game {
        // 2D array that holds whether a that spaces is taken or not on the game board
        pub board: [[Block; 22]; 10],
        pub next_board: [[[u8; 3]; 10]; 4],
        pub held_board: [[[u8; 3]; 4]; 4],

        // Tetriminos that are to be displayed in the game
        curr_mino: Mino,
        next_minos: [Mino; 3], // The next the 3 Tetriminos that will be dropped
        held_mino: Mino,

        // Boolean to know if there is a held mino
        has_held: bool,

        // Invisble componenets of Tetris
        level: usize,
        pub drop_ticks: u16, // Called speed but actually Milliseconds

        // Displayed values for the game
        pub lines: u16,
        pub time: u16, // Total Seconds
        pub score: u32,
    }

    impl Game {
        // Constructor
        pub fn new() -> Game {
            let mut tetris = Game {
                curr_mino : Mino::new(),
                next_minos : [Mino::new(), Mino::new(), Mino::new()],
                drop_ticks : DROP_TICKS[0],
                .. Default::default()
            };
            tetris.update_pos();
            tetris.update_preview();
            tetris
        }

        fn can_rotate(&self) -> bool {
            let piece = &self.curr_mino;

            match piece.shape {
                Shape::ShapeO => {
                    return false;
                }
                _ => {}
            }

            let cent_x = piece.center[0] as i16;
            let cent_y = piece.center[1] as i16;

            for i in 0..CORD_NUM {
                let curr_x = piece.cords[i][0] as i16;
                let curr_y = piece.cords[i][1] as i16;

                let new_x = cent_x + (curr_y - cent_y);
                let new_y = cent_y - (curr_x - cent_x);

                if new_x >= B_LEN as i16
                    || new_x < 0
                    || new_y < 0
                    || self.board[new_x as usize][new_y as usize].state == State::Placed
                {
                    return false;
                }
            }
            return true;
        }

        /*
         * This function is for the user, aka main, to call.
         * Technically, this could be done within the function
         * currently called "can_rotate". However, I find this way
         * to have a better naming convention that would be understandable
         * for myself and hopefully others, if they read this.
         */
        pub fn rotate(&mut self) {
            if self.can_rotate() {
                self.curr_mino.rotate();
                self.update_pos();
            }
        }

        fn can_shift(&self, dir: bool) -> bool {
            let mov: i8 = if dir { 1 } else { -1 };

            let piece = &self.curr_mino;

            let mut x = piece.center[0] as i8;
            let mut y = piece.center[1] as usize;

            let mut next_x = x + mov;

            if next_x >= B_LEN as i8
                || next_x < 0
                || self.board[next_x as usize][y].state == State::Placed
            {
                return false;
            }

            for i in 0..CORD_NUM {
                x = piece.cords[i][0] as i8;
                y = piece.cords[i][1] as usize;

                next_x = x + mov;

                if next_x >= B_LEN as i8
                    || next_x < 0
                    || self.board[next_x as usize][y].state == State::Placed
                {
                    return false;
                }
            }
            return true;
        }

        pub fn shift(&mut self, dir: bool) {
            if self.can_shift(dir) {
                self.curr_mino.shift(dir);
                self.update_pos();
            }
        }

        fn place(&mut self) {
            let mut x = self.curr_mino.center[0] as usize;
            let mut y = self.curr_mino.center[1] as usize;

            self.board[x][y].state = State::Placed;

            for i in 0..CORD_NUM {
                x = self.curr_mino.cords[i][0] as usize;
                y = self.curr_mino.cords[i][1] as usize;

                self.board[x][y].state = State::Placed;
            }
        }

        // Gets the distance the current Tetrimino will drop if dropped
        fn drop_distance(&self) -> usize {
            let calc_drop = |y: usize, board: &[Block]| -> usize {
                for j in (0..y).rev() {
                    if board[j].state == State::Placed {
                        return y - (j + 1);
                    }
                }
                return y;
            };

            let piece = &self.curr_mino;

            let cx = piece.center[0] as usize;
            let cy = piece.center[1] as usize;

            let mut drop = calc_drop(cy, &self.board[cx][0..cy]);

            for i in 0..CORD_NUM {
                let x = piece.cords[i][0] as usize;
                let y = piece.cords[i][1] as usize;

                if x != cx || y < cy {
                    let temp = calc_drop(y, &self.board[x][0..y]);
                    if temp < drop {
                        drop = temp
                    }
                }
            }
            return drop;
        }

        // The updates the location of the current tetrimino
        fn update_pos(&mut self) {
            let drop = self.drop_distance();
            let piece = &self.curr_mino;

            for i in 0..B_LEN {
                for j in 0..B_HEI {
                    match self.board[i][j].state {
                        State::Falling | State::Phantom => {
                            self.board[i][j].state = State::Empty;
                            self.board[i][j].color = COLOR_LIST[BLACK];
                        }
                        _ => {}
                    }
                }
            }

            let mut x = piece.center[0] as usize;
            let mut y = piece.center[1] as usize;

            self.board[x][y - drop].state = State::Phantom;
            self.board[x][y].state = State::Falling;
            self.board[x][y].color = COLOR_LIST[piece.shape as usize];
            self.board[x][y - drop].color = COLOR_LIST[piece.shape as usize];

            for i in 0..CORD_NUM {
                x = piece.cords[i][0] as usize;
                y = piece.cords[i][1] as usize;

                self.board[x][y - drop].state = State::Phantom;
                self.board[x][y].state = State::Falling;
                self.board[x][y].color = COLOR_LIST[piece.shape as usize];
                self.board[x][y - drop].color = COLOR_LIST[piece.shape as usize];
            }
        }

        // Updates held and next minos block positions
        fn update_preview(&mut self) {
            self.next_board = [[COLOR_LIST[BLACK]; 10]; 4];
            self.held_board = [[COLOR_LIST[BLACK]; 4]; 4];

            if self.has_held {
                let cent_x = 1;
                let cent_y = 1;

                self.held_board[cent_x][cent_y] = COLOR_LIST[self.held_mino.shape as usize];

                for i in 0..CORD_NUM {
                    let x =
                        (cent_x as i8 + CORD_LIST[self.held_mino.shape as usize][i][0]) as usize;
                    let y =
                        (cent_y as i8 + CORD_LIST[self.held_mino.shape as usize][i][1]) as usize;

                    self.held_board[x][y] = COLOR_LIST[self.held_mino.shape as usize];
                }
            }

            let cent_x = 1;
            let mut cent_y: isize = 7;
            for j in 0..NUM_MINOS {
                self.next_board[cent_x][cent_y as usize] =
                    COLOR_LIST[self.next_minos[j].shape as usize];

                for i in 0..CORD_NUM {
                    let x = (cent_x as i8 + CORD_LIST[self.next_minos[j].shape as usize][i][0])
                        as usize;
                    let y = (cent_y as i8 + CORD_LIST[self.next_minos[j].shape as usize][i][1])
                        as usize;

                    self.next_board[x][y] = COLOR_LIST[self.next_minos[j].shape as usize];
                }

                cent_y -= 3;
            }
        }

        // Releases the next mino making it the current Tetrimino, in the invisible
        // rows of the board (above row 20)
        fn release_next(&mut self) {
            self.curr_mino = self.next_minos[0];
            self.next_minos[0] = self.next_minos[1];
            self.next_minos[1] = self.next_minos[2];
            self.next_minos[2] = Mino::new();
            self.update_preview();
        }

        fn all_down_one(&mut self, bottom: usize) {
            for i in 0..B_LEN {
                for j in bottom..(B_HEI - 1) {
                    self.board[i][j] = self.board[i][j + 1];
                }
            }
        }

        // Reads the possible rows to clear and if they are full clears
        fn line_up(&mut self) {
            let mut num_rows = 0;
            let mut is_filled: bool;

            let mut j: usize = 0;
            while j < B_HEI {
                is_filled = true;
                for i in 0..B_LEN {
                    if self.board[i][j].state != State::Placed {
                        is_filled = false;
                        break;
                    }
                }
                j += 1;

                if is_filled {
                    j -= 1;
                    self.all_down_one(j);
                    self.lines += 1;
                    num_rows += 1;
                }

                if num_rows >= 4 {
                    break;
                }
            }

            if self.lines / (4 * (self.level + 1)) as u16 != 0 {
                self.level = min(MAX_LEVEL, self.level +1);
            }
        }

        fn can_fall_one(&self) -> bool {
            let piece = &self.curr_mino;

            let cx = piece.center[0] as usize;
            let cy = piece.center[1] as usize;

            if cy <= 0 || self.board[cx][cy - 1].state == State::Placed {
                return false;
            }

            for i in 0..CORD_NUM {
                let x = piece.cords[i][0] as usize;
                let y = piece.cords[i][1] as usize;

                if x == cx && y > cy {
                    continue;
                } else if y <= 0 || self.board[x][y - 1].state == State::Placed {
                    return false;
                }
            }
            return true;
        }

        pub fn fall_or_place(&mut self) {
            self.drop_ticks -= 1;
            if self.drop_ticks <= 0 {
                if self.can_fall_one() {
                    self.curr_mino.fall();
                } else {
                    self.place();
                    self.line_up();
                    self.release_next();
                }
                self.drop_ticks = DROP_TICKS[self.level];
                self.update_pos();
            }
        }

        // Drops the current Tetrimino on top of the pieces below it
        pub fn drop(&mut self) {
            let dist = self.drop_distance() as u8;
            let piece = &mut self.curr_mino;

            piece.center[1] -= dist;

            for i in 0..CORD_NUM {
                piece.cords[i][1] -= dist;
            }

            self.update_pos();
        }

        // Stores the current Tetrimino and releases the previously held Tetrimino
        // if there is one. Otherwise releases the next Tetrimino
        pub fn hold(&mut self) {
            let cent_x: i8 = 5;
            let cent_y: i8 = 20;

            if !self.has_held {
                self.held_mino = self.curr_mino;
                self.release_next();
                self.has_held = true;
            } else {
                let temp = self.curr_mino;
                self.curr_mino = self.held_mino;
                self.held_mino = temp;
            }

            let piece = &mut self.held_mino;

            piece.center[0] = cent_x as u8;
            piece.center[1] = cent_y as u8;

            for i in 0..CORD_NUM {
                piece.cords[i][0] = (cent_x + CORD_LIST[piece.shape as usize][i][0]) as u8;

                piece.cords[i][1] = (cent_y + CORD_LIST[piece.shape as usize][i][1]) as u8;
            }

            self.update_pos();
            self.update_preview();
        }
    }
}
